ocaml/xapi/message_forwarding.ml

```diff
@@ -3226,6 +3230,16 @@ module Forward = functor(Local: Custom_actions.CUSTOM_ACTIONS) -> struct
         with _ -> raise Not_found in
       do_op_on ~local_fn ~__context ~host op
 
+    (* Forward SR operation to all hosts that have suitable, plugged or
+       unplugged PBDs and the pool master *)
+    let forward_sr_all_op ~local_fn ~__context ~self op =
+      let module HostSet = Set.Make(struct type t = API.ref_host let compare = Pervasives.compare end) in
+      choose_pbds_for_sr ~consider_unplugged_pbds:true ~__context ~self ()
+      |> List.map (fun pbd -> Db.PBD.get_host ~__context ~self:pbd)
+      |> List.cons (Helpers.get_master ~__context)
```

Comment:
We are already on the master, so why would we forward the call to the master?

Reply:
I can try removing the master from the list, if it's present (that's the case of shared SRs) and directly perform the operation on master after all the calls are forwarderded. It will make the function have caveats and will be more difficult to use, however.

Reply:
I can try removing the master from the list, if it's present (that's the case of shared SRs) and directly perform the operation on master after all the calls are forwarderded. It will make the function have caveats and will be more difficult to use, however.

Reply:
Probably not worth it.

Reply:
I think it's needed to ensure the SR is not deleted from the database by the master; before the Database call in `destroy` happens in the SR master.

Where the operations are run in branch `master`:

| Op      | Shared | Not Shared (Master) | Not Shared (Slave)
| :------ | -----: | ------------------: | -----------------:
| Destroy | master | master              | slave
| Forget  | master | master              | master

Right now this branch doesn't behave like that. I'm changing xapi_sr destroy so it doesn't depend on forget and can accomplish the behaviour in the table. Then I'll adapt the selection mechanism based on that.

For reference, this is the desired behaviour:

| Op      | Shared | Not Shared (Master) | Not Shared (Slave)
| :------ | -----: | ------------------: | -----------------:
| Destroy | master | master              | slave
| Forget  | master | master              | master
| RRD rm  | all    | all                 | all

Reply:
Latest change removes the need to add the pool master to the list: if the SR is shared or if the SR is managed by the master it will already contain the reference to master. in the other case the RRD deletion does not need to be forwarded to it.

In any case if master is present the call will still be local because of the logic in `do_op_on`.